### DISCUSSION TITLE
v9.3 (2022/03)

### DISCUSSION THREAD

**Discussion by Mizux** [google/or-tools#3183] [Announces]
# Platform changes
* Drop Debian-10 support
* Drop Ubuntu-16.04 support
* Drop .NET Framework 4.5.2

# Dependencies Update
* Add Eigen `3.4.0`.
* Add Google re2 `2021-11-01`.
* Protobuf `3.19.1` -> `3.19.4`.
* SCIP `7.0.1` -> `v800`
## Python
* Add pybind11.

# Features
* Add PDLP as experimental
* Add MathOpt as experimental.

# CP-SAT
* renamed few API for consistency
 e.g. `LinearExpr.ScalProd` -> `LinearExpr.WeightedSum`.
* Add `AddAtLeastOne`/`AddAtMostOne`/`AddExactlyOne` methods.
* Add `AddMultiplicationConstraint(z, x, y)` in all languages.
* `AddMultipleCircuit()` in all languages

## C++
* explicit ctor `IntVar(BoolVar)`
* remove `LinearExpr::Add*`and replaced them by operator e.g. `LinearExpr += `
* Add arithmetic operators on linear expression
* removed `LinearExpr::BooleanSum/BooleanScalProd`, use `Sum/WeightedSum`
* Add `CpModelBuilder::FixVariable()` which overwrite the domain of the variable to a single value

## Java
* Rewrite `LinearExpr`, add a incremental builder class: `LinearExpr.newBuilder().add(x).addSum(<array of variables).build()`
* Follow C++ API: `Circuit`, `MultipleCircuit`, `Cumulative`, `Reservoir`, `AllowedAssignment` and `ForbiddenAssignment` now return a specialized class with incremental API to add new variables, terms, demands…

## C#
* Document all methods
* Follow C++ API: `Circuit`, `MultipleCircuit`, `Cumulative`, `Reservoir`, `AllowedAssignment` and `ForbiddenAssignment` now return a specialized class with incremental API to add new variables, terms, demands…
* Add `LinearExprBuilder` class to build expression incrementally.

# Build System
## CMake
* Require at least `CMake >= 3.18`
## Make
* Now use the CMake based build internally.

<hr /><em>This discussion was created from the release <a href='https://github.com/google/or-tools/releases/tag/v9.3'>v9.3 (2022/03)</a>.</em>

**Comment by imalliar** [comment]
Hi,
I would like to ask about dropping support for .NET 4.6  
Ortools 9.2 supported it ( https://www.nuget.org/packages/Google.OrTools/9.2.9972 ), but 9.3 isn't ( https://www.nuget.org/packages/Google.OrTools/9.3.10497 ) and the notes above does not mention this.
Note that 4.6.2 won't lost its support on April
Thanks in advance
Jacob

**Reply by lperron** [reply]
fixed the notes.

**Reply by Mizux** [reply]
Actually net452 was clunky.
because we only "support" it on windows (thanks to visual studio installing it) while on unix it was never supported (because you need mono etc...) BUT the 9.2 meta package nuspec contains:
```xml
 <dependencies>
      <group targetFramework=".NETFramework4.5.2">
        <dependency id="Google.OrTools.runtime.linux-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.osx-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.win-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.Protobuf" version="3.19.1" exclude="Build,Analyzers" />
      </group>
      <group targetFramework=".NETCoreApp3.1">
        <dependency id="Google.OrTools.runtime.linux-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.osx-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.win-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.Protobuf" version="3.19.1" exclude="Build,Analyzers" />
      </group>
      <group targetFramework="net6.0">
        <dependency id="Google.OrTools.runtime.linux-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.osx-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.win-x64" version="9.2.9972" exclude="Build,Analyzers" />
        <dependency id="Google.Protobuf" version="3.19.1" exclude="Build,Analyzers" />
      </group>
    </dependencies>
```
Which is **broken** by design since `Google.OrTools.runtime.osx-x64` and `Google.OrTools.runtime.linux-x64` don't support it (no subdirectory net452 with native libraries).

note: You can also build from source a custom version for you by modifying/hacking few lines:
in Meta package:
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/dotnet/Google.OrTools-local.csproj.in#L5
also removing this dead code
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/dotnet/Google.OrTools-local.csproj.in#L27-L33

Same for runtime template:
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/dotnet/Google.OrTools.runtime.csproj.in#L4

side note: 9.3 meta package nuspec contains
```xml
    <dependencies>
      <group targetFramework=".NETCoreApp3.1">
        <dependency id="Google.OrTools.runtime.linux-x64" version="9.3.10497" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.osx-x64" version="9.3.10497" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.win-x64" version="9.3.10497" exclude="Build,Analyzers" />
        <dependency id="Google.Protobuf" version="3.19.4" exclude="Build,Analyzers" />
      </group>
      <group targetFramework="net6.0">
        <dependency id="Google.OrTools.runtime.linux-x64" version="9.3.10497" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.osx-x64" version="9.3.10497" exclude="Build,Analyzers" />
        <dependency id="Google.OrTools.runtime.win-x64" version="9.3.10497" exclude="Build,Analyzers" />
        <dependency id="Google.Protobuf" version="3.19.4" exclude="Build,Analyzers" />
      </group>
    </dependencies>
```

**Reply by imalliar** [reply]
Thanks

**Comment by ktnr** [comment]
Thanks for the CMake rework! I don't know if I should open an issue for this. I'm on Windows.

When building or-tools form source, I get a `ortools.lib` and a `ortools_full.lib`. Now, my old programs only work if I link to `ortools_full.lib` instead of `ortools.lib`. 
With `ortools_full.lib` everything works fine. With `ortools.lib` I get a lot of linker errors of absl etc... What's the difference between the two? 

I also tried the binary release. It only ships an `ortools.lib`. With that my old programs do not work, I get multiple definition warnings like
```
1>ortools.lib(CbcCbcParam.obj) : warning LNK4006: "int CbcOrClpEnvironmentIndex" (?CbcOrClpEnvironmentIndex@@3HA) already defined in ortools.lib(CbcOrClpParam.obj); second definition ignored
```
and errors
```
1>ContainerLoading.lib(CbcCbcParam.obj) : error LNK2005: "int CbcOrClpEnvironmentIndex" (?CbcOrClpEnvironmentIndex@@3HA) already defined in ContainerLoading.lib(CbcOrClpParam.obj)
1>ContainerLoading.lib(CbcCbcParam.obj) : error LNK2005: "char * alternativeEnvironment" (?alternativeEnvironment@@3PEADEA) already defined in ContainerLoading.lib(CbcOrClpParam.obj)
1>C:\Users\s6362386\Source\Repos\ktnr\3L-VehicleRouting\ContainerLoading\x64\Release\ContainerLoadingApplication.exe : fatal error LNK1169: one or more multiply defined symbols found
```
Also, in the windows binary release the include folder is empty and it has a different folder structure than the build from source or the binary releases for linux.

**Reply by Mizux** [reply]
if you stick in the "CMake" realme you'll see cmake injecting all .lib to your project (e.g. cmake --build buid_dir -v on a C++ sample).
In Makefile to reproduce the behaviour before the cmake based build migration, and avoid any errors, we merge all .lib, like laurent said so currently we try to simulate this by merging all .lib.

What is your compile line ?
does ContainerLoading.lib a custom lib build by yourself ?

**Reply by Mizux** [reply]
note: To keep everything  in the same conversation I'm copying the following replies...

Laurent:
> at build time:
   or_tools_full contains all dependencies.
   or_tools.lib only the or-tools code.
   At archive time, we rename ortools_full.lib to ortools.lib

ktnr:
> Understood, thanks. Then I'm surprised that it doesn't work with the binary release, if it should be the same.

Laurent:
> especially since we do test the binary release :-)

**Reply by ktnr** [reply]
Dependencies are like this:
ContainerLoading.lib (library built by us)
- ortools.lib resp. ortools_full.lib
- ortools header files

ContainerLoadingApplication.exe (CLI app built by us)
- ContainerLoading.lib
- only ortools header files, no ortools.lib

The project is built with the standard VS build process with .vcxproj files, no CMake. 

When I build ortools from source with CMake and use it in the project, everything works fine. 

When I use the prebuilt binaries, ContainerLoading.lib raises the multiple definition warnings mentioned above (not the case when ortools is built from source) but the build completes without errors. However, when linking ContainerLoadingApplication and ContainerLoading, the above mentioned linker errors occur and the project cannot be built.

For completeness I provide the build and linker args.
<details>
<summary>ORTools Build</summary>

```
cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release -DBUILD_CXX=ON -DBUILD_PYTHON=OFF -DBUILD_DOTNET=OFF -DBUILD_DEPS=ON -DUSE_SCIP=OFF -DUSE_COINOR=OFF -DBUILD_SAMPLES=OFF -DBUILD_CXX_SAMPLES=OFF -DBUILD_PYTHON_SAMPLES=OFF -DBUILD_JAVA_SAMPLES=OFF -DBUILD_DOTNET_SAMPLES=OFF -DBUILD_EXAMPLES=OFF -DBUILD_CXX_EXAMPLES=OFF -DBUILD_PYTHON_EXAMPLES=OFF -DBUILD_JAVA_EXAMPLES=OFF -DBUILD_DOTNET_EXAMPLES=OFF -DCMAKE_INSTALL_PREFIX=install -DINSTALL_BUILD_DEPS=ON

cmake --build build --config Release --target ALL_BUILD

cmake --build build --config Release --target INSTALL
```
</details>

<details>
<summary>ContainerLoading.lib</summary>

- Build args:
```
/Zc:__cplusplus /permissive- /ifcOutput "x64\Release\" /GS /GL /W3 /wd"4996" /Gy /Zc:wchar_t /I"C:\Program Files\or-tools\include" 
/I"C:\gurobi950\win64/include" /Zi /Gm- /O2 /sdl /Fd"C:\Users\s6362386\Source\Repos\ktnr\3L- 
VehicleRouting\ContainerLoading\x64\Release\ContainerLoading.pdb" /Zc:inline /fp:precise /D "NDEBUG" /D "_LIB" /D "_UNICODE" /D 
"UNICODE" /errorReport:prompt /WX- /Zc:forScope /Gd /Oi /MD /std:c++latest /FC /Fa"x64\Release\" /EHsc /nologo /Fo"x64\Release\" 
/Fp"x64\Release\ContainerLoading.pch" /diagnostics:column
```

- Linker args:
```
/OUT:"C:\Users\s6362386\Source\Repos\ktnr\3L-VehicleRouting\ContainerLoading\x64\Release\ContainerLoading.lib" /LTCG 
"ortools_full.lib" "gurobi95.lib" "gurobi_c++md2019.lib" /MACHINE:X64 /NOLOGO /LIBPATH:"C:\Program Files\or-tools\lib" 
/LIBPATH:"C:\gurobi950\win64/lib"
```
</details>

<details>
<summary>ContainerLoadingApplication.exe</summary>

- Build args:
```
/Zc:__cplusplus /permissive- /ifcOutput "x64\Release\" /GS /GL /W3 /wd"4996" /Gy /Zc:wchar_t /I"C:\gurobi950\win64/include" /I"C:\Program Files\or- tools\include" /Zi /Gm- /O2 /sdl /Fd"x64\Release\vc143.pdb" /Zc:inline /fp:precise /D "NDEBUG" /D "_CONSOLE" /D "_UNICODE" /D 
"UNICODE" /errorReport:prompt /WX- /Zc:forScope /Gd /Oi /MD /std:c++latest /FC /Fa"x64\Release\" /EHsc /nologo /Fo"x64\Release\" 
/Fp"x64\Release\ContainerLoadingApplication.pch" /diagnostics:column
```

- Linker args:
```
/OUT:"C:\Users\s6362386\Source\Repos\ktnr\3L-VehicleRouting\ContainerLoading\x64\Release\ContainerLoadingApplication.exe" 
/MANIFEST /LTCG:incremental /NXCOMPAT /PDB:"C:\Users\s6362386\Source\Repos\ktnr\3L- VehicleRouting\ContainerLoading\x64\Release\ContainerLoadingApplication.pdb" /DYNAMICBASE "gurobi95.lib" 
"gurobi_c++md2019.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "comdlg32.lib" "advapi32.lib" "shell32.lib" "ole32.lib" 
"oleaut32.lib" "uuid.lib" "odbc32.lib" "odbccp32.lib" /DEBUG /MACHINE:X64 /OPT:REF /INCREMENTAL:NO 
/PGD:"C:\Users\s6362386\Source\Repos\ktnr\3L-VehicleRouting\ContainerLoading\x64\Release\ContainerLoadingApplication.pgd" 
/SUBSYSTEM:CONSOLE /MANIFESTUAC:"level='asInvoker' uiAccess='false'" 
/ManifestFile:"x64\Release\ContainerLoadingApplication.exe.intermediate.manifest" 
/LTCGOUT:"x64\Release\ContainerLoadingApplication.iobj" /OPT:ICF /ERRORREPORT:PROMPT 
/ILK:"x64\Release\ContainerLoadingApplication.ilk" /NOLOGO /LIBPATH:"C:\gurobi950\win64/lib" /TLBID:1
``` 
</details>

**Reply by Mizux** [reply]
```
1>ContainerLoading.lib(CbcCbcParam.obj) : error LNK2005: "int CbcOrClpEnvironmentIndex" (?CbcOrClpEnvironmentIndex@@3HA) already defined in ContainerLoading.lib(CbcOrClpParam.obj)
```

Does ContainerLoading source code depends on (i.e. include headers from) OR-Tools or Cbc ? in this case you should also have few defines
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/tools/Makefile.cc.java.dotnet#L131-L137
otherwise you may parse the header not like ortools thus conflict in two different definitions for the same symbol...

**Reply by ktnr** [reply]
Yes it does include OR-Tools headers but not Cbc dependencies. What do you mean by "have few defines" and what does it have to do with the makefile you linked? I was expecting there to be no difference in building ortools from source and then including it in the project (works) or using the prebuilt binaries and using it (does not work).

**Comment by jmarca** [comment]
In switching to 9.3, in C#, using CPSAT, how does one create constant IntVars with names?

I used to do:

```csharp
var constant = model.NewConstant(0,"do not track changes");
```
but that went away.  In C++, I have used `.WithName("blah blah")` but that doesn't seem to be exposed in c#.

Is is just impossible now to add names to constant IntVars?

**Comment by lperron** [comment]
I removed it. Do you really need it ? They were merged anyway in presolve.

**Comment by jmarca** [comment]
Well, it certainly isn’t a requirement, but the constants will be the only variables without names, which somewhat breaks my debugging code that dumps out the constraints. 

I guess I can create a regular IntVar, with a single value instead of a range, or perhaps with a constraint fixing the value. 

Are there optimizations that I will miss if I don’t use constant?

James

**Comment by lperron** [comment]
Or just an IntVar with a fixed domain.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mer. 23 mars 2022 à 16:16, James Marca ***@***.***> a
écrit :

> Well, it certainly isn’t a requirement, but the constants will be the only
> variables without names, which somewhat breaks my debugging code that dumps
> out the constraints.
>
> I guess I can create a regular IntVar, with a single value instead of a
> range, or perhaps with a constraint fixing the value.
>
> Are there optimizations that I will miss if I don’t use constant?
>
> James
>

**Reply by jmarca** [reply]
Yes I've switched to using that.  There are no optimizations missed from not using NewConstant(), correct?

**Comment by lperron** [comment]
Not really. Constants are merged during presolve. but you do not care.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mer. 23 mars 2022 à 17:14, James Marca ***@***.***> a
écrit :

**Comment by gregy4** [comment]
Some details about rewritten java layer for cp-sat ?
So far I miss any method with 'offset' in name like addLessOrEqualWithOffset for example. It is possible to rewrite it with LinearExpr.sum but it is very ugly to write it frequently.

old:
```java
   model.addEqualityWithOffset(taskStart, getTaskEndVariable(relation.getPreviousTask()), -relation.getMinDelta());
```
new:
```java
   model.addEquality(LinearExpr.sum(new LinearArgument[] {taskStart, LinearExpr.constant(-relation.getMinDelta())}),
                                    getTaskEndVariable(relation.getPreviousTask()));
```

I like most alternative with an operator in string like:
```java
   model.arithm(taskStart, "-", relation.getMinDelta(), "=", getTaskEndVariable(relation.getPreviousTask());
```

But it is my own taste.
Is it possible to make CpModel class not final ?  Writing my own alternatives would be easier and I can better survive changes in CpModel like current one.

**Comment by lperron** [comment]
We recommend using the builder paradigm.

LinearExpr.newBuilder().add(var).add(offset)

**Comment by LeiLi5** [comment]
Great work on the newly implemented PDLP. I'd like to ask about the next step.  Specifically any plans on a GPU version to accelerate PDLP?
Thank you in advance.

**Reply by LeiLi5** [reply]
I would be excited if that's on your plate.
